### DISCUSSION TITLE
Modifying costing profile by disallowing edges after GetBestPath is run

### DISCUSSION THREAD

**Discussion by kungfu303** [valhalla/valhalla#4422] [Q&A]
I have a project where I’d like to run GetBestPath() over multiple source and destination combinations. If the shortest path is within a certain maximum distance, then I would like to add the source and destination edges to a running list. The goal is then to eliminate those edges in the running list from the graph after each iteration of GetBestPath(). My initial idea is to disallow access to them in the Allowed() function of the costing profile I send to GetBestPath().

Where would you suggest I write this within Valhalla? I was thinking of doing it within the test directory, but unsure how to expose the list of eliminated edges to the costing profile that GetBestPath() uses (for example, bicyclecost.cc). I would probably need to add an extra parameter to GetBestPath or the costing profile to allow for this list.

**Comment by nilsnolde** [comment]
TBH I don't really understand what it is you're trying to do. I mean what the intention is. Can you try rephrasing or give an example? E.g. "eliminate those edges in the running list from the graph" is very confusing. On what basis do you want to eliminate what? It's impossible to remove anything from the graph, so I guess you mean some edges shouldn't be considered anymore in some next iteration of the routing algo? Why more iterations, to do what? Public Valhalla only ever does 2 iterations, and that's only if the first one failed. Somehow it sounds like a crude description of "alternative paths" but since that's already there, probably not?

Sounds like you have smth very special in mind. Are you planning to contribute smth upstream? Giving custom free dev consultation for proprietary projects would be a bit beyond the scope of a volunteer-run OSS project:)

**Comment by kevinkreiser** [comment]
I think hes saying he will call GetBestPath himself in his own code, the library has an api and there is nothing stopping him from doing that. What it seems like he'd like is to then on subsequent calls disallow edge that were used in previous calls.

We have a way to do this using the costing. Currently we have avoid_locations which allows you to specify in a regular request locations that you dont want the route using. Internally how that works is that it finds edges near those locations and adds them to a blacklist inside of the costing so that its `allowed` method says those edges are not allowed when the algorithm encounters them.

So after a run of `GetBestPath`, you can take the edges you're interested in removing and add them to this blacklist that the costing manages. see here:

https://github.com/valhalla/valhalla/blob/ffe790ddfe08b89c02c419ec08cfec0bf715a64c/valhalla/sif/dynamiccost.h#L820

Now if you are saying you want to do all of this via the standard http API that is tough.. The reason that is tough is because we dont expose a direct way for you to disable edges by edge_id in the http API. Technically you could send tons of `avoid_locations` but that would be only good to do a proof of concept. To make this usable you'd need to extend both the route response to include edge ids somehow (if its map matching that is done already) and then a way to allow as input the same edge ids. This also opens you up to the ids changing mid request stream if for example your service gets new data while you are in the middle of this string of calls to the API. what i mean is the edge_ids are not stable from version to version of the routing data.
